drivers/gpio/gpio_si32_hd.c

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) 2023 GARDENA GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/gpio/gpio_utils.h>
+#include <zephyr/dt-bindings/gpio/silabs-si32-hd-gpio.h>
+#include <zephyr/kernel.h>
+
+#include <SI32_PBCFG_A_Type.h>
+#include <SI32_PBHD_A_Type.h>
+#include <si32_device.h>
+
+struct gpio_si32_hd_config {
+	struct gpio_driver_config common;
+	SI32_PBHD_A_Type *base;
+
+	uint32_t nchannel_current_limit;
+	bool has_nchannel_current_limit;
+
+	uint32_t pchannel_current_limit;
+	bool has_pchannel_current_limit;
+
+	bool enable_bias;
+	bool low_power_port_mode;
+	bool enable_drivers;
+};
+
+struct gpio_si32_hd_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+};
+
+static int gpio_si32_hd_configure(const struct device *dev, gpio_pin_t pin, gpio_flags_t flags)
+{
+	const struct gpio_si32_hd_config *config = dev->config;
+
+	/* Set pin function to GPIO.
+	 * This is the only supported mode for now.
+	 */
+	switch (pin) {
```

Comment:
This entire switch statement looks like it should be reduced to 1 line of code. Is there really no function that takes the pin as a parameter?

Reply:
There isn't because we only use the register-descriptions of the HAL so there's not much abstraction or simplification on top of what the registers look like. We also can't just calculate the offset within the `PBHD4_PBFSEL` register based on the pin number because pin5 is special and uses a different number of bits.

Reply:
You could use the `FOR_EACH` macro from Zephyr to avoid this code duplication.